{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "114f0449-bf20-482b-9701-7f3ef5b853e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Accession                 68\n",
       "Organism_Name             68\n",
       "GenBank_RefSeq            68\n",
       "Assembly                  68\n",
       "Nucleotide                68\n",
       "Submitters                67\n",
       "Release_Date              68\n",
       "Isolate                   36\n",
       "Species                   68\n",
       "Length                    68\n",
       "Protein                   68\n",
       "Geo_Location              57\n",
       "USA                        3\n",
       "Host                      55\n",
       "Tissue_Specimen_Source     9\n",
       "Collection_Date           49\n",
       "dtype: int64"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#finding common strains\n",
    "import pandas as pd\n",
    "\n",
    "#load files\n",
    "spike = pd.read_csv(\"spike.csv\", sep=\",\")\n",
    "nsp12 = pd.read_csv(\"nsp12.csv\", sep=\",\")\n",
    "nuc = pd.read_csv(\"nucleocapsid.csv\", sep=\",\")\n",
    "\n",
    "spike.count()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "1a1a1202-c1d9-423c-8a17-95326d072dd2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 15 organism names\n"
     ]
    }
   ],
   "source": [
    "#extract organism name\n",
    "spike_names = set(spike['Organism_Name']) \n",
    "nsp12_names = set(nsp12['Organism_Name'])\n",
    "nuc_names = set(nuc['Organism_Name'])\n",
    "\n",
    "#shared organism name\n",
    "shared = spike_names & nsp12_names & nuc_names\n",
    "\n",
    "print(\"Found\", len(shared), \"organism names\") \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "55c4cc9a-f0c3-4a34-8b02-048f7a9c1948",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 16 organism names\n"
     ]
    }
   ],
   "source": [
    "shared_spike_nsp12 = spike_names & nsp12_names\n",
    "print(\"Found\", len(shared_spike_nsp12), \"organism names\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "bc678949-5711-4e8a-a85a-f8ba1ee62a06",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 17 organism names\n"
     ]
    }
   ],
   "source": [
    "shared_nsp12_nuc = nsp12_names & nuc_names\n",
    "print(\"Found\", len(shared_nsp12_nuc), \"organism names\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "fcc74d93-da1c-4faf-9a24-9679403b7c77",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 57 organism names\n"
     ]
    }
   ],
   "source": [
    "shared_nuc_spike = spike_names & nuc_names\n",
    "print(\"Found\", len(shared_nuc_spike), \"organism names\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "9758eacd-c6fe-4525-84db-03551dfa27e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 60 assemblies\n"
     ]
    }
   ],
   "source": [
    "#check with assembly\n",
    "spike_assembly = set(spike['Assembly']) \n",
    "nsp12_assembly = set(nsp12['Assembly']) \n",
    "nuc_assembly = set(nuc['Assembly']) \n",
    "\n",
    "shared_nuc_spike_assembly = spike_assembly & nuc_assembly\n",
    "print(\"Found\", len(shared_nuc_spike_assembly), \"assemblies\") \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "094f630a-0879-44ee-91ed-4a0abb21cf93",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'GCF_001500975.1', 'GCF_000886515.1', 'GCF_000896875.1', 'GCF_023141795.1', 'GCF_001503155.1', 'GCF_023148835.1', 'GCF_000879255.1', 'GCF_012271615.1', 'GCF_001661775.1', 'GCF_000895415.1', 'GCF_012271625.1', 'GCF_000873025.1', 'GCF_012271565.1', 'GCF_002194405.1', 'GCF_002118885.1', 'GCF_001962315.1', 'GCF_000856025.1', 'GCF_000887595.1', 'GCF_000848685.1', 'GCF_000853865.1', 'GCF_000926915.1', 'GCF_003971785.1', 'GCF_001505415.1', 'GCF_012271635.1', 'GCF_001501755.1', 'GCF_000930095.1', 'GCF_012271745.1', 'GCF_023124615.1', 'GCF_000894435.1', 'GCF_000870505.1', 'GCF_002816235.1', 'GCF_000864885.1', 'GCF_000870985.1', 'GCF_000868165.1', 'GCF_002816175.1', 'GCF_001504755.1', 'GCF_000896895.1', 'GCF_000862965.1', 'GCF_012271575.1', 'GCF_002816215.1', 'GCF_002816195.1', 'GCF_000896035.1', 'GCF_009858895.2', 'GCF_000879875.1', 'GCF_000853505.1', 'GCF_000896935.1', 'GCF_001725835.1', 'GCF_000880835.1', 'GCF_000875645.1', 'GCF_023122875.1', 'GCF_000919475.1', 'GCF_000901155.1', 'GCF_000883335.1', 'GCF_000899495.1', 'GCF_000858765.1', 'GCF_000880055.1', 'GCF_000868045.1', 'GCF_000872845.1', 'GCF_012271735.1', 'GCF_002219865.1'}\n"
     ]
    }
   ],
   "source": [
    "print(shared_nuc_spike_assembly)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "469f5e11-38ee-468d-b7b1-1047df214270",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "60\n",
      "61\n"
     ]
    }
   ],
   "source": [
    "#filter rows with shared assemblies\n",
    "\n",
    "spike_filt = spike[spike['Assembly'].isin(shared_nuc_spike_assembly)] \n",
    "nuc_filt = nuc[nuc['Assembly'].isin(shared_nuc_spike_assembly)] \n",
    "\n",
    "print(spike_filt['Assembly'].count())\n",
    "print(nuc_filt['Assembly'].count())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "53443fe8-a9e3-459c-935f-619bd7e0e75b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    YP_010037478.1\n",
      "1    YP_010037564.1\n",
      "2    YP_010115720.1\n",
      "3    YP_010799347.1\n",
      "4    YP_009825061.1\n",
      "Name: Accession, dtype: object\n"
     ]
    }
   ],
   "source": [
    "#extract accession ids\n",
    "\n",
    "spike_acc = spike_filt['Accession']\n",
    "nuc_acc = nuc_filt['Accession']\n",
    "\n",
    "print(nuc_acc.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "8f1f0168-4a3c-4446-a479-3d5be41988e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#extract sequences with accession ids\n",
    "\n",
    "#nucleocapsid\n",
    "keep = False\n",
    "with open(\"nucleocapsid.fasta\") as f_in, open (\"nucleocapsid_filtered.fasta\", \"w\") as f_out: \n",
    "    for line in f_in:\n",
    "        if line.startswith(\">\"):\n",
    "            acc_id = line[1:].strip().split(\" \")[0]\n",
    "            if acc_id in set(nuc_acc): \n",
    "                keep = True\n",
    "                f_out.write(line)\n",
    "            else: \n",
    "                keep = False\n",
    "        else: \n",
    "            if keep == True: \n",
    "                f_out.write(line) \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "e07558bc-ce44-4f8b-92e3-454204197980",
   "metadata": {},
   "outputs": [],
   "source": [
    "#spike\n",
    "with open(\"spike.fasta\") as f_in, open (\"spike_filtered.fasta\", \"w\") as f_out: \n",
    "    for line in f_in:\n",
    "        if line.startswith(\">\"):\n",
    "            acc_id = line[1:].strip().split(\" \")[0]\n",
    "            if acc_id in set(spike_acc): \n",
    "                keep = True\n",
    "                f_out.write(line)\n",
    "            else: \n",
    "                keep = False\n",
    "        else: \n",
    "            if keep == True: \n",
    "                f_out.write(line) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "771b7b73-b131-401a-86d8-0971e578d21c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#rewrite headers of aligned fasta files\n",
    "\n",
    "#nucleocapsid\n",
    "with open(\"nucleocapsid_aligned.fasta\") as aln_file, open(\"nucleocapsid_aln_clean.fasta\", \"w\") as clean_file:\n",
    "    for line in aln_file: \n",
    "        if line.startswith(\">\"):\n",
    "            acc = line[1:].strip().split(\" \")[0]\n",
    "            gene = line.split(\"|\")[1].split(\"[\")[0].strip().replace(\" \", \"_\")\n",
    "            org_name = line.strip().split(\"[\")[1].split(\"]\")[0].replace(\" \", \"_\")\n",
    "            new_head = f\">{org_name} | {acc} | {gene}\"\n",
    "            clean_file.write(new_head + \"\\n\") \n",
    "        else: \n",
    "            clean_file.write(line) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e04439d7-bce3-4b39-9fdf-8cf449b2d770",
   "metadata": {},
   "outputs": [],
   "source": [
    "#repeat for spike\n",
    "\n",
    "with open(\"spike_aligned.fasta\") as aln_file, open(\"spike_aln_clean.fasta\", \"w\") as clean_file:\n",
    "    for line in aln_file: \n",
    "        if line.startswith(\">\"):\n",
    "            acc = line[1:].strip().split(\" \")[0]\n",
    "            gene = line.split(\"|\")[1].split(\"[\")[0].strip().replace(\" \", \"_\")\n",
    "            org_name = line.strip().split(\"[\")[1].split(\"]\")[0].replace(\" \", \"_\")\n",
    "            new_head = f\">{org_name} | {acc} | {gene}\"\n",
    "            clean_file.write(new_head + \"\\n\") \n",
    "        else: \n",
    "            clean_file.write(line)  "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
